### PR TITLE
CXP-1245: refactor criterion registry

### PR DESCRIPTION
**Pull request by tseho**
**Description (for Contributor and Core Developer)**

We are introducing criterion on attributes with CXP-1245, after some digging, I found out that the code will be cleaner if we first refactor the existing CriterionRegistry.

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/catalogs/front/src/components/ProductSelection/hooks/useFindCriterionByField.unit.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+const criterions: string[] = ['enabled', 'family'];
+
+test.each(criterions)('it returns a criterion when searching for "%s"', async field => {
+    const {result} = renderHook(() => useFindCriterionByField());
+
+    expect(await result.current(field)).toMatchObject({
+        component: expect.any(Function),
+        factory: expect.any(Function),
+    });
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/hooks/useFindCriterionByField.unit.ts:14]: I find this test a bit lazy. Can't we test if the return object is of the correct type ?

File: components/catalogs/front/src/components/ProductSelection/hooks/useSystemCriterionFactories.unit.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    expect(result.current).toEqual([
+        {
+            label: 'akeneo_catalogs.product_selection.criteria.status.label',
+            factory: expect.any(Function),
+        },
+        {
+            label: 'akeneo_catalogs.product_selection.criteria.family.label',
+            factory: expect.any(Function),
+        },
+    ]);
```
Comment (masacc) [inline components/catalogs/front/src/components/ProductSelection/hooks/useSystemCriterionFactories.unit.ts:18]: Is this test really useful?

**Review by masacc — APPROVED** [review]

**Review by RiffVentura — APPROVED** [review]

File: components/catalogs/front/src/components/ProductSelection/hooks/useSystemCriterionFactories.unit.ts (hunk 2 above)
Comment (tseho) [reply components/catalogs/front/src/components/ProductSelection/hooks/useSystemCriterionFactories.unit.ts:18]: Not much but required, we still have 100% coverage requirement on hooks.

File: components/catalogs/front/src/components/ProductSelection/hooks/useFindCriterionByField.unit.ts (hunk 1 above)
Comment (tseho) [reply components/catalogs/front/src/components/ProductSelection/hooks/useFindCriterionByField.unit.ts]: Better ? :)

Comment (masacc) [reply components/catalogs/front/src/components/ProductSelection/hooks/useFindCriterionByField.unit.ts]: Yey ! \o/

**Pull request merged**